### ISSUE TITLE
Eliminate node creation when viewing the FRBR structure

### ISSUE THREAD

**Issue by patrickzurek** [eXtensibleCatalog/Drupal-Toolkit#305]
JIRA issue created by: pkiraly
Originally opened: 2011-07-14 10:47 AM

Issue body:
 (nt)

**Comment by patrickzurek** [comment]
JIRA Comment by user: pkiraly
JIRA Timestamp: 2011-07-14 10:47 AM

Comment body:

For historical reason, which is unnecessary now, Drupal Toolkit creates node for Work, Expression and Holding records. This is not necessary, and even confuse the statistics. I should rewrite the function.

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2012-03-08 07:03 PM

Comment body:

As I understand it, viewing the XC Schema button creates nodes for Works and Expressions (not sure about holdings). Peter opened this issue to discontinue this, but then I think Dave has an interest in getting nodes created for all.

I also understand that Drupal Toolkit only used the Manifestation nodes.

Please advise on how to proceed with this issue, including, leave it as it is.

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2012-06-19 10:21 AM

Comment body:

Please advise on this issue.

**Comment by patrickzurek** [comment]
JIRA Comment by user: mwesley
JIRA Timestamp: 2012-06-19 05:39 PM

Comment body:

I think this is similar to issue #DRUPAL-112 and #DRUPAL-241 in respect to the way we store metadata in Drupal. With that, I would like to have Dave's [~dlindahl] opinion on this whole topic, as well as Randy's opinion and everyone else. 

The following is a summary of the historical reasons Peter mentioned. Initially, we wanted to have nodes for everything, because that would be the most flexible and "Drupal" way of doing things. It would allow us or anyone to use many Drupal modules to plug into our metadata management. 

Now, this seems a bit overkill, especially in respect to harvesting. With harvesting, it ends up that the node-creation process is the slowest process. In addition, the node-deletion process is also very slow. This is because, the "correct" method requires handling each node separately and running multiple hooks for it. To speed up these slow processes, we've created a lot of tricks for performance sake so that now nodes are not really useful anymore. Even the Manifestation nodes are not necessarily very useful. I assume this is why Peter reduced the node creation to only Manifestation records, because they are the only one's used heavily. We use nodes for (1) some few external modules, such as the printer friendly pages, and (2) to view a full record after search -- however, even in this case, the node's view is not of the Manifestation record itself, or even the node's information from Drupal's SQL database, but the view is created with metadata from the combined Work-Manifestation-Expression-Holding information stored in Solr. This is why search functionality can work even without any nodes created.

Basically, I think we need to rethink whether we need to have nodes at all, or at least if we want any nodes to be created automatically. Perhaps we can just make sure we have the same full record view and give admins the ability to convert metadata into nodes if they really want to. That would save us a lot of time with the post-harvest processes, reduce some Drupal overhead, and make it easier for us to do metadata additions and deletions without SQL tricks.

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2012-06-20 10:03 AM

Comment body:

Assigning this to Dave for input.

**Comment by patrickzurek** [comment]
JIRA Comment by user: dlindahl
JIRA Timestamp: 2012-06-20 11:43 AM

Comment body:

If we are not using the nodes for anything, let's get rid of them.  
1) How will eliminating them affect things that use them currently, like you mentioned printing?
2) How much time will it take (programming time) to eliminate the use of the nodes in the code? 
3) Are we talking about eliminating node creation entirely, or is there some tradeoff relating to programmer time versus benefit of removing all the nodes?
4) How do we currently support developers who want to write XC add-ons that use the data in our XC records? I assume they can just go into SOLR? (I am not looking for long answers to any of these questions - yes or no would suffice in some cases)
5) Is there any "future-use" reason that you can think of to create nodes?  If so, can we preserve the code we are removing for future re-adding?
6) We always envisioned XC being able to search across website content AND XC schema records in the same search. One way to do that is to have all the data in Drupal Nodes, another is to put it all in SOLR, and the third is to have some in SOLR and some in Drupal Nodes. I suspect that the only practical way to do it (without writing our own search engine) is to index website data in SOLR together with the XC schema records.  Is this correct?

**Comment by patrickzurek** [comment]
JIRA Comment by user: rcook
JIRA Timestamp: 2012-06-20 11:55 AM

Comment body:

As MT prepares a response I wanted to frame this.   The issue was started because we do create Manifestion nodes, but not other nodes (W, E, etc.) unless someone views the XC Schema view.   That leaves us with a mixed bag of nodes, but as we do not use those W and E nodes, I had proposed that we turn them off.  I do not think we need to remove the creation of M nodes at this point (we do use them, even though it sounds like with reprograming we could remove their use), unless they are problematic in some way.  I guess I am an advocate of turning off what we don't use, but not spend time to rework what is working (unless there is a reason to).

**Comment by patrickzurek** [comment]
JIRA Comment by user: mwesley
JIRA Timestamp: 2012-06-26 10:00 AM

Comment body:

As Randy said, this issue is really related to an issue with the "View XC Schema" link. For that, I've committed a fix for the node creation. It was mostly unnecessary, but just needed a rewrite. Now, the properties load without creating nodes, by only using the information in the metadata entity. So, that should close this issue.

**Comment by patrickzurek** [comment]
JIRA Comment by user: mwesley
JIRA Timestamp: 2012-06-26 10:37 AM

Comment body:

1) How will eliminating them affect things that use them currently, like you mentioned printing?

I don't think the print functionality is actually working and I'm not aware of any contributed modules that use our metadata nodes -- we don't even use the node much. When someone views the node page, the Solr document is retrieved for that Manifestation, which includes metadata from other entities related to that Manifestation. This process is someway wrong in my opinion, but right now it may be the fastest and easiest way to get all the information needed to display a full record page.

2) How much time will it take (programming time) to eliminate the use of the nodes in the code? 

It would take about a week. I think it is worth it, but there are other important things to work on now, such as being able to stop a scheduled harvest that is running but stalls or being able to restart the harvest from where it left off, especially if run from cron. 

3) Are we talking about eliminating node creation entirely, or is there some tradeoff relating to programmer time versus benefit of removing all the nodes?

Yes, perhaps. We are talking about either (1) eliminating automatic node creation or (2) eliminating it all together. If a developer was tapping into our modules, there would be no difference between using nodes and not using nodes. However, if an admin of a library's website using Drupal wanted to use an existing module with our system (for example, using the 5-star module to allow content rating) they would want metadata to be representable as nodes. 

4) How do we currently support developers who want to write XC add-ons that use the data in our XC records? I assume they can just go into SOLR? (I am not looking for long answers to any of these questions - yes or no would suffice in some cases)

There is no real support for what developers would want, such as a more usable API, although it is very much possible to plug into the XC modules right now. Without nodes, they could still tap into the Metadata or Solr to retrieve metadata information. 

5) Is there any "future-use" reason that you can think of to create nodes? If so, can we preserve the code we are removing for future re-adding?

I think we should allow admins to create nodes afterwards if desired. This is already somewhat implemented, since you can select to skip node-generation after harvesting and you can also manually start node-generation. Also, all our code is in the Git repository.

6) We always envisioned XC being able to search across website content AND XC schema records in the same search. One way to do that is to have all the data in Drupal Nodes, another is to put it all in SOLR, and the third is two have some in SOLR and some in Drupal Nodes. I suspect that the only practical way to do it (without writing our own search engine) is to index website data in SOLR together with the XC schema records. Is this correct?

This was always a bit confusing to me. I think there were talks about putting website content _into_ XC schema, which I don't know if it's possible or right to do. If instead, we're indexing website content with Solr, then that's not difficult and in fact is already done with the Apache Solr search module for Drupal. If we wanted to do so as well, our indexing and search modules can allow this to happen in part. However, much of the search interface is customized for XC schema alone, and moreover, Manifestation records alone. I don't know how long it would take to make the code more generic, or even if we want to, since the work we have done so far that has been more generic and flexible has also been less performant and more cumbersome to understand.

**Comment by patrickzurek** [comment]
JIRA Comment by user: mwesley
JIRA Timestamp: 2012-06-26 07:38 PM

Comment body:

The work has been completed. I think the conversation about removing nodes entirely can be moved to another issue.

**Comment by patrickzurek** [comment]
Issue resolved: 2012-06-26 07:38 PM
